distributed/stealing.py

```diff
@@ -417,4 +425,62 @@ def story(self, *keys):
         return out
 
 
+def _can_steal(ws, what, from_):
```

Comment:
So this is going to be called relatively often. AFAICT, it'll be called once per task, thief-worker pair for every task we're considering stealing. Given that, we should ensure that it's fast. How confident are you that there aren't any surprising performance issues here? Most things look OK, but may want to look into `get_address_host`.

- `get_address_host`

Reply:
In general, `get_address_host` is linear in the length of the input address, however I think it's safe to assume that `len(ws.address)` is small and bounded as a function of `ws`. 

The four collections used here are 
1. `what.host_restrictions`,
2. `what.worker_restrictions`,
3. `ws.resources`, and
4. `from_.resources`.

Key containment is checked using (1) and (2), and since they are sets these operations are constant with respect to the size of the containers. (3) and (4) are used in `_has_resources`. The work in the loop only uses (3). Because (3) is a dict, value retrieval is constant with respect to the size of (3). The loop iterates over (4) once, so the total work in this function is linear in the size of (4). So overall, `_can_steal` is O(`len(from_.resources)`).

Reply:
Yeah, I'm also concerned about the performance here.  

This code gets called *often*.  Previously we didn't iterate through `idle`, we just indexed into it (indeed we made it a `SortedSet` object just to get this constant-time access). 

The work stealing code accounts for a non-trivial amount of the scheduler performance costs.  I would not be surprised if adding this check significantly increases those costs.  Given that resources aren't used in the common case I'm very hesitant to go this route without strong demonstration that it doesn't affect performance.

*However*, I would be much more comfortable with this if it was only engaged in the uncommon case that the task had resource constraints.  It could be that a simple test like the following resolves the issue.

```python
if ts.resources:
    theives = [...]
    theif = ...
else:
    thief = self.idle[i * len(self.idle)]
```

Then this PR is essentially a no-op in the common case, and we can be a lot more relaxed about performance.